sklearn/datasets/openml.py

```diff
@@ -0,0 +1,415 @@
+            warn('target_column_name={} has flag is_row_identifier.'.format(
+                target_column_name))
+    if len(found_types) != 1:
+        raise ValueError('Can only handle homogeneous multi-target datasets, '
+                         'i.e., all targets are either numeric or '
+                         'categorical.')
+    return list(found_types)[0]
+
+
+def fetch_openml(name=None, version='active', data_id=None, data_home=None,
+                 target_column_name='default-target', cache=True):
+    """Fetch dataset from openml by name or dataset id.
+
+    Datasets are uniquely identified by either an integer ID or by a
+    combination of name and version (i.e. there might be multiple
+    versions of the 'iris' dataset). Please give either name or id
+    (not both). In case a name is given, a version can also be
+    provided.
+
+    Parameters
+    ----------
+    name : str or None
+        String identifier of the dataset. Note that OpenML can have multiple
+        datasets with the same name.
+
+    version : integer or 'active', default='active'
+        Version of the dataset. Can only be provided if also ``name`` is given.
+        If 'active' the oldest version that's still active is used.
+
+    data_id : int or None
+        OpenML ID of the dataset. The most specific way of retrieving a
+        dataset. If data_id is not given, name (and potential version) are
+        used to obtain a dataset.
+
+    data_home : string or None, default None
+        Specify another download and cache folder for the data sets. By default
+        all scikit-learn data is stored in '~/scikit_learn_data' subfolders.
+
+    target_column_name : string, list or None, default 'default-target'
+        Specify the column name in the data to use as target. If
+        'default-target', the standard target column a stored on the server
+        is used. If ``None``, all columns are returned as data and the
+        target is ``None``. If list (of strings), all columns with these names
+        are returned as multi-target (Note: not all scikit-learn classifiers
+        can handle all types of multi-output combinations)
+
+    cache : boolean, default=True
+        Whether to cache downloaded datasets using joblib.
+
+    Returns
+    -------
+
+    data : Bunch
+        Dictionary-like object, the interesting attributes are:
+        'data', the data to learn, 'target', the regression target or
+        classification labels, 'DESCR', the full description of the dataset,
+        'feature_names', the original names of the dataset columns, and
+        'details' which provide more information on the openml meta-data.
+    """
+    data_home = get_data_home(data_home=data_home)
+    data_home = join(data_home, 'openml')
+    if cache:
+        mem = Memory(join(data_home, 'cache'), verbose=0).cache
+    else:
+        def mem(func):
+            return func
+    cached_get_data_info_by_name = mem(_get_data_info_by_name)
```

Comment:
Can we avoid cache invalidation in the case that any of these functions or their line in the file changes?? Is there a way to make the impact of changing this file have the least impact on the most problematic cache invalidation?

I suggested use of joblib.Memory here, but it also means that datasets can't be shared across virtual environments, unless I'm much mistaken, again because the function being defined in a different file is enough to invalidate the cache.

We might, therefore, need a custom storage implementation, using `joblib.dump` but not `joblib.Memory`.

Reply:
I feel like storing in a compact format and being able to read from any environment are slightly at odds, at least if we take the naive approaches. We could store the ARFF and it would be perfectly readable from any environment and basically no changes to the code would break anything. It takes a while to store it, though.
We can use joblib.dump but as you said that's a bit fragile as witnessed by the unvendoring joblib attempt. We could potentially consider caching in another generic file format if there is an efficient one that we have an efficient reader for.
hdf5 would be the natural choice I think but a reader/writer would be an extra dependency. I'm not sure how fast/efficient netcdf is, which is in scipy.

If we want to get something out quickly (which we do) maybe storing the ARFF is the easiest for the RC since we know it won't break and we could "upgrade" in the future?

Reply:
If we added pandas as a dependency we could use hdf5, parquet or feather ;)

Reply:
then again, this discussion should probably be inside openml and we shouldn't worry too much?

Reply:
> If we added pandas as a dependency we could use hdf5, parquet or feather ;)

Each of which would require to add one fairly complex extra dependency (independently of pandas) with a benefit that is not clear: loading the datasets from disk is probably not that much of a bottleneck in scikit-learn..

Note sure about hdf5 in scipy, but if we can put a data in the form of a array + metadata we might as well use `joblib.dump`. It's just a cache, I don't think we should worry to much about inter-version compatibility or long therm archival. Maybe some mechanism for invalidating the cache (e.g. if the pickle version changes) would be good, but we can also add this in future version if this is currently blocking for the release.

If one has to re-download the datasets for each scikit-learn release, I wouldn't say it's such a big deal.

> I suggested use of joblib.Memory here, but it also means that datasets can't be shared across virtual environments, unless I'm much mistaken, again because the function being defined in a different file is enough to invalidate the cache.

That's seems very unfortunate. No way around it in joblib?

Reply:
Not intending to rush anyone, but AFAIC there's no conclusion yet here?

Reply:
@janvanrijn none yet

Reply:
@rth I'm totally ok with just using it for caching and re-downloading for now as well. I didn't realize pandas required the other libraries as dependencies but it makes total sense.

I think a bigger problem is it breaking with some weird error message after scikit-learn upgrade. We had that in the past and worked around it, though.

Reply:
I think we can use joblib... But we should maybe check with joblib what kind of pickle stability they promise...

Reply:
I suspect for large ARFF the overhead of parsing it and interpreting types will be the most substantial cost to dataset loading.